### PR TITLE
Error handling for scan of external table.

### PR DESCRIPTION
**Pull request by Hlinbit**
In the task of scanning the external table, we will encounter some wrong data, but the error handling mechanism of gpdb can ensure that these errors will not affect the normal progress of the task.

But if we want to further calculate and process the scanned data, some exceptions that cannot be handled will be generated, and the task will fail. Here I take three examples：

> 1. division by zero
> 2. invalid input syntax for type inet: "12.3 4.56.78"
> 3. unsupported Unicode escape sequence

First, If the data in the foreign table is the following, we use the SQL `select 6/(column-1) from ext_table;`  to calculate the value of the given formula. The task will cause the "division by zero" error. 
```
  column
----------
      1
      2
      3
```

Another example is as follows. The data type is JSON, and each row is a formatted JSON whose key is 'ip' and the value is an address. However, the third row contains an invalid address value. When we want to get the value part of the JSON through the SQL `SELECT (c->>'ip')::inet FROM ext_table;`, an error 'invalid input syntax for type' inet' occurred.
```
             column
----------------------
{"ip": "157.230.122.5"}
{"ip": "214.55.78.31"}
{"ip": "54.    198.104.230"}
```

The final example is as follows. The data type is JSON, and each row is a formatted JSON whose key is a number and the value is a string. The first row contains some unsupported Unicode '\\u0000'. When we want to get the value part of the JSON through the SQL `SELECT (a->>'0') FROM ext_table;`, an error 'unsupported Unicode escape sequence' occurred.

```
                                                    column
----------------------------------------------------------------------------
{"0": "\\u0000賖\\u0000\\u0000\\u0000\\u0000\\u0000\\u0000\\u0000鼣"}
{"1": "yes"}
```

All the above tasks will be aborted by the corresponding error. We can look up the query plan by the `explain` command. And the plan is as follows. All the above errors occur during the `Foreign Scan` phase of the query plan. Thus, we call such errors foreign scan errors.

```
                                      QUERY PLAN
--------------------------------------------------------------------------------------
 Gather Motion 2:1  (slice1; segments: 2)  (cost=0.00..46000.00 rows=2000000 width=4)
   ->  Foreign Scan on ext_table  (cost=0.00..16000.00 rows=1000000 width=4)
 Optimizer: Postgres query optimizer
(3 rows)` 
``` 

However, such errors are not fatal. If gpdb can handle these errors and record them into a log file for the users to check,  it will improve the user experience when it comes to some simple processing of external data.

So this PR is for handling the projection errors that have not been handled by gpdb currently.  In the code, most of the exceptions generated during the foreign scan phase are handled except for some specific errors. 

The `SEGMENT REJECT LIMIT` clause allows you to isolate format errors in external table data and to continue loading correctly formatted rows in the original gpdb. In this modification, the clause is also applied to handle the foreign scan error. Use `SEGMENT REJECT LIMIT` to set an error threshold, specifying the reject limit count as the number of ROWS (the default) or as a `PERCENT` of total rows (1-100).  And this threshold includes all foreign scan errors including format errors.

These foreign scan errors are uniformly collected into files, and these errors can be viewed through the SQL function `gp_read_error_log`.

To implement the feature, we add the exception handling mechanism in the execution callback function `ExecForeignScan`. If an exception is thrown in the call stack of `ExecForeignScan`, it will be caught. The exception should be judged on whether it can be handled when `SEGMENT REJECT LIMIT` is set. The exception that is able to be handled will be written into a file for review through `gp_read_error_log`. 

In test cases for the modification, we prove the effectiveness for division of zero, JSON mapping error, and unsupported Unicode error.

### CODE REVIEW

File: src/backend/executor/nodeForeignscan.c [hunk 1]
```diff
@@ -112,9 +112,25 @@ ExecForeignScan(PlanState *pstate)
 {
 	ForeignScanState *node = castNode(ForeignScanState, pstate);
 
-	return ExecScan(&node->ss,
-					(ExecScanAccessMtd) ForeignNext,
-					(ExecScanRecheckMtd) ForeignRecheck);
+	TupleTableSlot *result = NULL;
+	int get_error = 0;
```
Comment (hyongtao-db) [inline src/backend/executor/nodeForeignscan.c:116]: Why `Int`? Is `Boolean `better here?

**Review by hyongtao-db — COMMENTED** [review]
Code is neat and good. Thanks for your great contribution.

This pr can help FDW handles tolerable but not fatal errors， take these 3 example：
- division by zero
- invalid input syntax for type inet: "12.3    4.56.78"
- unsupported Unicode escape sequence

I have 2 tiny suggestions:
#### First one
Could you please modify the commit log to make it more readable？Take an example below:
###### Example 1：division by zero
concrete content
###### Example 2：division by zero
concrete content
###### Example 3：division by zero
concrete content

#### Second one
Could you please add some other test case like below:
`errmsg("value \"%s\" is out of range for type %s", str, "money")));`
No need to cover all this errmsg in test case. But you had better to make a simple summary in commit log.

Thanks again.

File: src/backend/executor/nodeForeignscan.c [hunk 2]
```diff
@@ -112,9 +112,25 @@ ExecForeignScan(PlanState *pstate)
...
+	do 
+	{
+		PG_TRY();
+		{
```
Comment (zhaorui20005) [inline src/backend/executor/nodeForeignscan.c:120]: Need to reset got_error here, otherwise an infinite loop may be triggered if the last line of data has error.

File: src/include/foreign/fdwapi.h [hunk 3]
```diff
@@ -200,6 +202,7 @@ typedef struct FdwRoutine
 	IterateForeignScan_function IterateForeignScan;
 	ReScanForeignScan_function ReScanForeignScan;
 	EndForeignScan_function EndForeignScan;
+	ForeignScanErrorHandle_function ErrorHandle;
```
Comment (zhaorui20005) [inline src/include/foreign/fdwapi.h:205]: ScanErrorHandle is a better name.

File: src/test/regress/output/external_table.source [hunk 4]
```diff
@@ -5004,3 +5004,100 @@ SELECT logerrors from pg_exttable a, pg_class b where a.reloid = b.oid and b.rel
...
+location ('file://localhost.localdomain/home/gpadmin/workspace/gpdb/src/test/regress/data/eh_divzero.dat')
+FORMAT 'text' (delimiter '|') LOG ERRORS SEGMENT REJECT LIMIT 2;
+SELECT 6/(c-2) FROM ext_divzero;
+ERROR:  segment reject limit reached, aborting operation  (seg0 slice1 127.0.0.1:7000 pid=16238)
+DETAIL:  Last error was: division by zero
+SELECT count(*) FROM gp_read_error_log('ext_divzero');
+ count 
+-------
+     2
+(1 row)
```
Comment (zhaorui20005) [inline src/test/regress/output/external_table.source:5018]: Is the line of data which causes the projection error recorded? What does the buffer `cstate->line_buf.data;` contain when projection error happened? Considering both "text|csv" and custom formatter.

File: src/backend/commands/copy.c [hunk 5]
```diff
@@ -5268,7 +5268,7 @@ NextCopyFrom(CopyState cstate, ExprContext *econtext,
  *
  * changing me? take a look at FILEAM_HANDLE_ERROR in fileam.c as well.
  */
-static void
+void
 HandleCopyError(CopyState cstate)
 {
 	if (cstate->errMode == ALL_OR_NOTHING)
```
Comment (zhaorui20005) [inline src/backend/commands/copy.c:5274]: I am curious that how does `HandleCopyError` handle errors like divide 0 error, because it only handles `ERRCODE_TO_CATEGORY(elog_geterrcode()) == ERRCODE_DATA_EXCEPTION` error code.

File: src/backend/executor/nodeForeignscan.c [hunk 6]
```diff
@@ -397,3 +413,12 @@ ExecShutdownForeignScan(ForeignScanState *node)
 	if (fdwroutine->ShutdownForeignScan)
 		fdwroutine->ShutdownForeignScan(node);
 }
+
+void
+ExecForeignScanError(ForeignScanState *node)
+{
+	FdwRoutine *fdwroutine = node->fdwroutine;
+	if (fdwroutine->ErrorHandle)
+		fdwroutine->ErrorHandle(node);
```
Comment (zhaorui20005) [inline src/backend/executor/nodeForeignscan.c:422]: If not define this handler, this function should continue throwing out.

File: src/test/regress/output/external_table.source (hunk 4 above)
Comment (Hlinbit) [reply src/test/regress/output/external_table.source:5018]: Yes,
 the content of the error line will be recorded.

File: src/backend/commands/copy.c (hunk 5 above)
Comment (Hlinbit) [reply src/backend/commands/copy.c:5274]: OK because the error 'division of zero' is one case in  ERRCODE_DATA_EXCEPTION. And the macro ERRCODE_TO_CATEGORY can get the category of the error code.

File: src/test/regress/output/external_table.source (hunk 4 above)
Comment (zhaorui20005) [reply src/test/regress/output/external_table.source:5018]: Then show them in the testing cases no only count(*), and there is no custom formatter cases.

File: src/backend/cdb/cdbsreh.c [hunk 7]
```diff
@@ -301,7 +301,7 @@ ReportSrehResults(CdbSreh *cdbsreh, uint64 total_rejected)
 	if (total_rejected > 0)
 	{
 		ereport(NOTICE,
-				(errmsg("found " INT64_FORMAT " data formatting errors (" INT64_FORMAT " or more input rows), rejected related input data",
+				(errmsg("found " INT64_FORMAT " foreign scan errors (" INT64_FORMAT " or more input rows), rejected related input data",
```
Comment (zhaorui20005) [inline src/backend/cdb/cdbsreh.c:304]: No need to change.

File: src/test/regress/output/external_table.source (hunk 4 above)
Comment (Hlinbit) [reply src/test/regress/output/external_table.source:5018]: Displaying content is a non-achievable feature because of the timestamp.

-               cmdtime               |        relname        |                                                                                  filename                                                                                   | linenum | bytenum |                                errmsg                                 | rawdata | rawbytes 
--------------------------------------+-----------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+---------+-----------------------------------------------------------------------+---------+----------
- Fri Apr 14 01:52:02.133033 2023 PDT | ext_fixed_width_table | file://localhost.localdomain//home/gpadmin/workspace/gpdb/src/test/regress/data/eh_fixed_width.dat [//home/gpadmin/workspace/gpdb/src/test/regress/data/eh_fixed_width.dat] |       5 |         | expected line size from the formatting string is 22, actual size is 0 |         | \x
+              cmdtime               |        relname        |                                                                                  filename                                                                                   | linenum | bytenum |                                errmsg                                 | rawdata | rawbytes 
+------------------------------------+-----------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+---------+-----------------------------------------------------------------------+---------+----------
+ Fri Apr 14 01:53:13.54256 2023 PDT | ext_fixed_width_table | file://localhost.localdomain//home/gpadmin/workspace/gpdb/src/test/regress/data/eh_fixed_width.dat [//home/gpadmin/workspace/gpdb/src/test/regress/data/eh_fixed_width.dat] |       5 |         | expected line size from the formatting string is 22, actual size is 0 |         | \x
 (1 row)

Comment (Hlinbit) [reply src/test/regress/output/external_table.source:5018]: However, for the custom formatter, the raw data is not available for gp_read_error_log, while the line number and all other columns in the error log are available.

File: src/include/foreign/fdwapi.h [hunk 8]
```diff
@@ -200,6 +202,7 @@ typedef struct FdwRoutine
 	IterateForeignScan_function IterateForeignScan;
 	ReScanForeignScan_function ReScanForeignScan;
 	EndForeignScan_function EndForeignScan;
+	ForeignScanErrorHandle_function ScanErrorHandle;
```
Comment (zhaorui20005) [inline src/include/foreign/fdwapi.h:205]: ScanErrorHandler

File: src/test/regress/input/external_table.source [hunk 9]
```diff
@@ -3667,3 +3667,80 @@ DROP EXTERNAL TABLE ext_false;
...
+
+DROP EXTERNAL TABLE ext_badjson;
+
+CREATE EXTERNAL TABLE ext_badjson(a json) 
+location ('file://localhost.localdomain/home/gpadmin/workspace/gpdb/src/test/regress/data/eh_badjson.dat') 
+FORMAT 'text' LOG ERRORS SEGMENT REJECT LIMIT 10;
+
+SELECT (a->>'9') FROM ext_badjson;
+
+SELECT errmsg FROM gp_read_error_log('ext_badjson');
```
Comment (zhaorui20005) [inline src/test/regress/input/external_table.source:3725]: The most useful information inside error log is what is the origin data line, and maybe the line number and filename. I think you can use `select * from ` for simplicity and then we can see much clearer the whole information about the error message. 
You can refer to this doc <https://docs.vmware.com/en/VMware-Tanzu-Greenplum/6/greenplum-database/GUID-admin_guide-load-topics-g-viewing-bad-rows-in-the-error-table-or-error-log.html> for help.

File: gpcontrib/gp_exttable_fdw/extaccess.c [hunk 10]
```diff
@@ -752,6 +753,28 @@ external_insert(ExternalInsertDesc extInsertDesc, TupleTableSlot *slot)
 	pstate->fe_msgbuf->data[0] = '\0';
 }
 
+static void
+copy_sinfo_from_string(CopyState pstate, char *src, int len)
+{
+	char *data = NULL;
+
+	/*  */
+	while ( src[len - 1] == '\n' || src[len - 1] == '\r' )
```
Comment (huiliang-liu) [inline gpcontrib/gp_exttable_fdw/extaccess.c:762]: we should check `len` to make sure `len-1` is valid

File: gpcontrib/gp_exttable_fdw/extaccess.c [hunk 11]
```diff
@@ -752,6 +753,28 @@ external_insert(ExternalInsertDesc extInsertDesc, TupleTableSlot *slot)
...
+	if (len > 0)
+	{
+		if(pstate->need_transcoding) 
+		{
+			data = pg_any_to_server(src, len, pstate->file_encoding);
+			if (data != src) {
+				len = strlen(data);
+			}
+			pstate->line_buf_converted = true;
+		}
```
Comment (huiliang-liu) [inline gpcontrib/gp_exttable_fdw/extaccess.c:774]: should set `data` as `src` if `pstate->need_transcoding` is false/null

File: gpcontrib/gp_exttable_fdw/extaccess.c [hunk 12]
```diff
@@ -752,6 +753,28 @@ external_insert(ExternalInsertDesc extInsertDesc, TupleTableSlot *slot)
...
+
+	/*  */
+	while ( src[len - 1] == '\n' || src[len - 1] == '\r' )
+		len--;
+
+	if (len > 0)
+	{
+		if(pstate->need_transcoding) 
+		{
+			data = pg_any_to_server(src, len, pstate->file_encoding);
```
Comment (huiliang-liu) [inline gpcontrib/gp_exttable_fdw/extaccess.c:769]: why decoding the raw data?

File: gpcontrib/gp_exttable_fdw/gp_exttable_fdw.c [hunk 13]
```diff
@@ -810,6 +810,13 @@ static void exttable_EndForeignInsert(EState *estate,
 	external_insert_finish(extInsertDesc);
 }
 
+static void
+exttable_ErrorHandling(ForeignScanState *node)
+{
+	exttable_fdw_state *fdw_state = (exttable_fdw_state *) node->fdw_state;
+	extTableErrorHandling(fdw_state->ess_ScanDesc);
```
Comment (huiliang-liu) [inline gpcontrib/gp_exttable_fdw/gp_exttable_fdw.c:817]: so, it seems that we needn't add `extTableErrorHandling` function? it is only used here

File: gpcontrib/gp_exttable_fdw/extaccess.c (hunk 12 above)
Comment (Hlinbit) [reply gpcontrib/gp_exttable_fdw/extaccess.c:769]: The function is to ensure the encoding rule is consistent with the master segment. After calling the `pg_any_to_server` we set the `line_buf_converted` to be `true`. And the raw data of the error line could be written into the file.

File: src/backend/commands/copy.c [hunk 14]
```diff
@@ -5369,9 +5369,22 @@ HandleCopyError(CopyState cstate)
 		 * if the error count is reached.
 		 */
 		cdbsreh->rawdata->cursor = 0;
-		cdbsreh->rawdata->data = cstate->line_buf.data;
-		cdbsreh->rawdata->len = cstate->line_buf.len;
-
+		if (cstate->line_buf_converted == false) {
```
[the remark this answers is not in the document]
Comment (Hlinbit) [reply src/backend/commands/copy.c:5372]: In the data retrieval phase, information from 'text' and 'csv' formats is transformed to be compatible with the master segment. However, the 'custom' format consists of a byte sequence, making it difficult to determine if the data can be converted. If unconverted data is detected during error handling, it is essential to attempt converting the byte data into a format compatible with the master segment. Otherwise, the error log file will only display the raw byte data.

Why is there only a sequence of bytes in the 'custom' format?
Various binary data types, such as 'json_header' and 'delimiter_header,' are used to store additional information that guides the parsing process. As a result, gpdb can only obtain a data sequence, which is then fed into the 'formatter' to generate the corresponding tuples.

[the remark this answers is not in the document]
Comment (zhaorui20005) [reply src/backend/commands/copy.c:5372]: That's fine to convert text and csv format row data. But from the view of customers who are using custom formatter, they need the origin data to debug or to trace the error data. So I think we should not tailor the eol in row data, and there is no need to make the decoding especially when the row data is json or binary. The PR's mission is to record the row data but not decoding or handle them.

File: gpcontrib/gp_exttable_fdw/extaccess.c [hunk 15]
```diff
@@ -984,6 +987,24 @@ externalgettup_custom(FileScanDesc scan)
...
+			 * start and end positions of a line. Unfortunately, cursor operations may 
+			 * not be reliable across all formatters. If the cursor remains unchanged, 
+			 * GPDB will store all remaining data in fmt_databuf into line_buf.
+			 */
+
+			int line_len = formatter->fmt_databuf.cursor - line_begin;
+			if ( line_len == 0 && 
+				formatter->fmt_databuf.len > formatter->fmt_databuf.cursor)
+			{
+				line_len = formatter->fmt_databuf.len - line_begin;
```
Comment (zhaorui20005) [inline gpcontrib/gp_exttable_fdw/extaccess.c:1006]: Can you explain why need to set line_len to `formatter->fmt_databuf.len - line_begin` when line_len equals to 0?

File: gpcontrib/gp_exttable_fdw/extaccess.c [hunk 16]
```diff
@@ -990,18 +1008,27 @@ externalgettup_custom(FileScanDesc scan)
...
 			{
 				switch (formatter->fmt_notification)
 				{
 					case FMT_NONE:
 
 						/* got a tuple back */
-
 						tuple = formatter->fmt_tuple;
-
+
```
[the remark this answers is not in the document]
Comment (zhaorui20005) [reply gpcontrib/gp_exttable_fdw/extaccess.c:1022]: Need to remove empty line.

File: src/backend/utils/mb/mbutils.c [hunk 17]
```diff
@@ -577,6 +577,7 @@ pg_client_to_server(const char *s, int len)
 	return pg_any_to_server(s, len, ClientEncoding->encoding);
 }
 
+
```
Comment (zhaorui20005) [inline src/backend/utils/mb/mbutils.c:580]: Remove unrelated empty line.

File: gpcontrib/gp_exttable_fdw/extaccess.c (hunk 15 above)
Comment (Hlinbit) [reply gpcontrib/gp_exttable_fdw/extaccess.c:1006]: For instance, when dealing with Avro data, the cursor will not update until all the data in fmt_databuf has been processed. If an error occurs during the avro processing, the line_len will not reflect the error data range, and we need to record all data in fmt_databuf.

Comment (zhaorui20005) [reply gpcontrib/gp_exttable_fdw/extaccess.c:1006]: When avro returns a tuple, will the line_len still be 0?

Comment (Hlinbit) [reply gpcontrib/gp_exttable_fdw/extaccess.c:1006]: Yes.

Comment (Hlinbit) [reply gpcontrib/gp_exttable_fdw/extaccess.c:1006]: Line_len is calculated by cursor minus the last cursor value. So if the cursor is not updated, the line_len equals 0.

Comment (zhaorui20005) [reply gpcontrib/gp_exttable_fdw/extaccess.c:1006]: So when will avro update the cursor?After returning all the tuples or after returning each tuple or after finishing parsing one line and returning several tuples?

Comment (Hlinbit) [reply gpcontrib/gp_exttable_fdw/extaccess.c:1006]: The cursor will be updated when Avro message reaches EOF. In the middle process, tuples will be returned continuously

Comment (zhaorui20005) [reply gpcontrib/gp_exttable_fdw/extaccess.c:1006]: Then there is no need to make this change for the avro, because the whole file content will be recorded.

Comment (Hlinbit) [reply gpcontrib/gp_exttable_fdw/extaccess.c:1006]: The content should be added into pstate->line_buf. The start and end of the data should be provided. If we want to the code meet the cursor updated and the cursor not updated, we have to check the condition in 'if'.

Comment (Hlinbit) [reply gpcontrib/gp_exttable_fdw/extaccess.c:1006]: And not only avro, but alse json, muticontainer have the same problem like avro.

Comment (zhaorui20005) [reply gpcontrib/gp_exttable_fdw/extaccess.c:1006]: I did not read the code of avro and json, but I have read multicontainer, it is not true for multicontainer. For multicontainer, the cursor will be added at end of each message, but not the whole file content until EOF. And for json, I remembered there maybe multiline-json, the cursor should be added at each end of message.
And in your implementation and assumption, if filter error happens, then it will record from the start of the file to the buffer that loaded at last loop.

Comment (Hlinbit) [reply gpcontrib/gp_exttable_fdw/extaccess.c:1006]: Do you mean that there is the possibility that the same content will be appended repeatedly?

Comment (zhaorui20005) [reply gpcontrib/gp_exttable_fdw/extaccess.c:1006]: In the scenario that if one formatter that does not add `fmt_databuf.cursor` until seeing the EOF, the fmt_databuf will be appended continuously which may cause a lot of memory which is not a good action. Then it is the formatter's fault we do not need to handle. Even in the way you have implemented, the recorded result is meaningless.

File: gpcontrib/gp_exttable_fdw/extaccess.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (zhaorui20005) [reply gpcontrib/gp_exttable_fdw/extaccess.c:1017]: Please remove this unrelated line with tabs.

File: src/backend/executor/nodeForeignscan.c [hunk 18]
```diff
@@ -112,9 +112,26 @@ ExecForeignScan(PlanState *pstate)
...
+							(ExecScanAccessMtd) ForeignNext,
+							(ExecScanRecheckMtd) ForeignRecheck);
+		}
+		PG_CATCH();
+		{
+			ExecForeignScanError(node);
+			get_error = 1;
+		}
+		PG_END_TRY();
+	} while (!result && get_error);
```
Comment (adam8157) [inline src/backend/executor/nodeForeignscan.c:132]: In which case it will throw the error and return a NULL result?

Seems this loop only returns good results, there is no code handling case `!result && !get_error`.

File: src/backend/executor/nodeForeignscan.c [hunk 19]
```diff
@@ -112,9 +112,26 @@ ExecForeignScan(PlanState *pstate)
 {
 	ForeignScanState *node = castNode(ForeignScanState, pstate);
 
-	return ExecScan(&node->ss,
-					(ExecScanAccessMtd) ForeignNext,
-					(ExecScanRecheckMtd) ForeignRecheck);
+	TupleTableSlot *result = NULL;
+	int get_error = 0;
```
Comment (adam8157) [inline src/backend/executor/nodeForeignscan.c:116]: `bool error_caught`

File: gpcontrib/gp_exttable_fdw/gp_exttable_fdw.c [hunk 20]
```diff
@@ -822,6 +830,7 @@ gp_exttable_fdw_handler(PG_FUNCTION_ARGS)
 	routine->IterateForeignScan = exttable_IterateForeignScan;
 	routine->ReScanForeignScan = exttable_ReScanForeignScan;
 	routine->EndForeignScan = exttable_EndForeignScan;
+	routine->ScanErrorHandler = exttable_ErrorHandling;
```
Comment (adam8157) [inline gpcontrib/gp_exttable_fdw/gp_exttable_fdw.c:833]: Other handlers all start with verbs, I propose

```
routine->HandleForeignScanError = exttable_HandleForeignScanError;
```

Other places are likewise.

File: src/backend/executor/nodeForeignscan.c (hunk 18 above)
Comment (beeender) [reply src/backend/executor/nodeForeignscan.c:132]: if `get_error = 1`, `result` must be `NULL` I think? Why does the while loop is needed?

Comment (Hlinbit) [reply src/backend/executor/nodeForeignscan.c:132]: If an error occurs in ExecScan, 'result' is null. However, if there is no more data, the result is also to be null. Thus, the loop makes sure that a null result will be returned only if the scan is over.

Comment (Hlinbit) [reply src/backend/executor/nodeForeignscan.c:132]: According to the function calling chain, the function 'execMotionSender' will get the result. If the result is null, the function will call 'doSendEndOfStream' and the transaction will be ended. Thus, only in the case that the scan is over, the null result could be returned.

**Review by zhaorui20005 — APPROVED** [review]
LGTM

File: gpcontrib/gp_exttable_fdw/extaccess.c [hunk 21]
```diff
@@ -995,13 +1011,19 @@ externalgettup_custom(FileScanDesc scan)
...
-
 						if (pstate->cdbsreh)
+						{
 							pstate->cdbsreh->processed++;
+							resetStringInfo(&pstate->line_buf);
+							/* 
+							 * Cursor is the start of next row, and row_size is the size of last row. 
+							 * So the start of last row is cursor minus row_size.
+							 */
+							char *src =  formatter->fmt_databuf.data + line_begin;
```
Comment (adam8157) [inline gpcontrib/gp_exttable_fdw/extaccess.c:1024]: Nitpick, redundant space character. Even the src variable is redundant, it's only used once.

**Review by adam8157 — COMMENTED** [review]
Could you add a test case that the error is not a format issue or one of the three issues you mentioned? Does it exist?

File: gpcontrib/gp_exttable_fdw/extaccess.c [hunk 22]
```diff
@@ -995,13 +1011,20 @@ externalgettup_custom(FileScanDesc scan)
...
-
 						tuple = formatter->fmt_tuple;
-
 						if (pstate->cdbsreh)
+						{
 							pstate->cdbsreh->processed++;
+							resetStringInfo(&pstate->line_buf);
+							/* 
+							 * Cursor is the start of next row, and row_size is the size of last row. 
+							 * So the start of last row is cursor minus row_size.
```
Comment (huiliang-liu) [inline gpcontrib/gp_exttable_fdw/extaccess.c:1022]: It seems that the comment doesn't match the code.

**Review by adam8157 — APPROVED** [review]

**Review by huiliang-liu — APPROVED** [review]

**Comment by Hlinbit** [comment]
In the task of scanning the external table, we will encounter some wrong data, but the error handling mechanism of gpdb can ensure that these errors will not affect the normal progress of the task.

But if we want to further calculate and process the scanned data, some exceptions that cannot be handled will be generated, and the task will fail. Here I take three examples：

> 1. division by zero
> 2. invalid input syntax for type inet: "12.3 4.56.78"
> 3. unsupported Unicode escape sequence

First, If the data in the foreign table is the following, we use the SQL `select 6/(column-1) from ext_table;`  to calculate the value of the given formula. The task will cause the "division by zero" error. 
```
  column
----------
      1
      2
      3
```

Another example is as follows. The data type is JSON, and each row is a formatted JSON whose key is 'ip' and the value is an address. However, the third row contains an invalid address value. When we want to get the value part of the JSON through the SQL `SELECT (c->>'ip')::inet FROM ext_table;`, an error 'invalid input syntax for type' inet' occurred.
```
             column
----------------------
{"ip": "157.230.122.5"}
{"ip": "214.55.78.31"}
{"ip": "54.    198.104.230"}
```

The final example is as follows. The data type is JSON, and each row is a formatted JSON whose key is a number and the value is a string. The first row contains some unsupported Unicode '\\u0000'. When we want to get the value part of the JSON through the SQL `SELECT (a->>'0') FROM ext_table;`, an error 'unsupported Unicode escape sequence' occurred.

```
                                                    column
----------------------------------------------------------------------------
{"0": "\\u0000賖\\u0000\\u0000\\u0000\\u0000\\u0000\\u0000\\u0000鼣"}
{"1": "yes"}
```


All the above tasks will be aborted by the corresponding error. We can look up the query plan by the `explain` command. And the plan is as follows. All the above errors occur during the `Foreign Scan` phase of the query plan. Thus, we call such errors foreign scan errors.

```
                                      QUERY PLAN
--------------------------------------------------------------------------------------
 Gather Motion 2:1  (slice1; segments: 2)  (cost=0.00..46000.00 rows=2000000 width=4)
   ->  Foreign Scan on ext_table  (cost=0.00..16000.00 rows=1000000 width=4)
 Optimizer: Postgres query optimizer
(3 rows)` 
``` 


However, such errors are not fatal. If gpdb can handle these errors and record them into a log file for the users to check,  it will improve the user experience when it comes to some simple processing of external data.

So this PR is for handling the projection errors that have not been handled by gpdb currently.  In the code, most of the exceptions generated during the foreign scan phase are handled except for some specific errors. 

The `SEGMENT REJECT LIMIT` clause allows you to isolate format errors in external table data and to continue loading correctly formatted rows in the original gpdb. In this modification, the clause is also applied to handle the foreign scan error. Use `SEGMENT REJECT LIMIT` to set an error threshold, specifying the reject limit count as the number of ROWS (the default) or as a `PERCENT` of total rows (1-100).  And this threshold includes all foreign scan errors including format errors.

These foreign scan errors are uniformly collected into files, and these errors can be viewed through the SQL function `gp_read_error_log`.

To implement the feature, we add the exception handling mechanism in the execution callback function `ExecForeignScan`. If an exception is thrown in the call stack of `ExecForeignScan`, it will be caught. The exception should be judged on whether it can be handled when `SEGMENT REJECT LIMIT` is set. The exception that is able to be handled will be written into a file for review through `gp_read_error_log`. 

In test cases for the modification, we prove the effectiveness for division of zero, JSON mapping error, and unsupported Unicode error.
